text/0000-custom-model-classes.md

```diff
@@ -0,0 +1,242 @@
+any registered callback for the given identifier, and pass in the type of the notification, allowing the record the opportunity to repull the data if needed. There are no guarantees around the timing of the notification callback being called after `RecordData` informs the store of changes. We expect that in a modern Ember app with tracked properties, this wouldn't be the common path for tracking changes. 
+
+```ts
+function unsubscribe(token: UnsubscribeToken)
+
+interface NotificationCallback {
+  (identifier: Identifier, notificationType: 'attributes' | 'relationships' | 'errors'  | 'meta'  | 'unload'): void;
+}
+    
+interface NotificationManager {
+  subscribe(identifier, NotificationCallback): UnsubscribeToken
+ }
+```
+
+## Exposing schema information
+
+We currently keep schema information on `DS.Model` class. In order to allow for custom Model implementations we need to allow lookup of schema info from the store. We already have specified a schema api that RecordData consumes: `attributesDefinitionFor` and `relationshipDefinitionFor`. We would define on a schema interface that the store would expose and addons could use. **The schema methods are not ergonomic on purpose.** They match the current Record Data apis and are designed as a stepping stone on the way of having a better, user facing schema APIs. Addons could provide their own `SchemaDefinitionService` by calling `registerSchemaDefinitionService`. We would initially not allow calling `registerSchemaDefinitionService` more than once, but this constraint could potentially be relaxed in the future. The schema info is currently primarily geared towards being used by the internal relationship handling layer, the serializer/adapter layers and the DebugAdapter. Schema methods support both static and dynamic schema computation. For static schemas, the method can always respond with a schema definition based on the type passed, and for dynamic changing schemas, it can look up the underlying data by the identifer which is passed in. We would also add a method called `isModelType`, which would return `true` if ED knew that a given string is a model type and `false` otherwise.
+  
+
+```ts
+export interface RelationshipDefinition {
+  kind: 'hasMany'| 'belongsTo';
+  type: string;
+  options: { [key: string]: any } ;
+  name: string;
+}
+    
+interface RelationshipsDefinition {
+  [key: string]: RelationshipDefinition
+}
+    
+interface AttributeDefinition {
+  name: string;
+  options: { [key: string]: any };
+  type: string;
+}
+
+interface AttributesDefinition {
+  [key: string]: AttributeDefinition
+}
+    
+interface SchemaDefinitionService {
+  // Following the existing RD implementation 
+  attributesDefinitonFor(identifier: RecordIdentifier | type: string): AttributesDefiniton
+  
+  // Following the existing RD implementation
+  relationshipsDefinitionFor(identifier: RecordIdentifier | type: string): RelationshipsDefinition
+
+  isModelType(type: string): boolean
```

Comment:
I think `isModelType` is pretty awkward. Something along the lines of `doesTypeExist` or `typeExists` seems a little more clear (imo).